### PR TITLE
custom marshaler: handle `Accept` headers correctly

### PR DESCRIPTION
**Pull request by tamird**
Also greatly simplifies existing logic to avoid special casing of
defaults.

### CODE REVIEW

**Comment by tamird** [comment]
@yugui @willtrking @tmc

**Comment by tamird** [comment]
This failed on `go tip` because `go vet` crashed, but is otherwise green.

**Comment by tmc** [comment]
why the MakeMarshalerMIMERegistry naming change?

**Comment by tamird** [comment]
Because it no longer returns a pointer.

**Comment by yugui** [comment]
I have another similar cleanup and bug fixes in #151.
I'll merge this PR into #151.

**Comment by yugui** [comment]
It is unlikely that I will merge this PR because
* I have a similar cleanup #151.
* I don't think this is a "proper" handling of Accept header since it ignores media range and quality factor of the header.  https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.1
* Accept header is not the right way to identify MIME type of request body.

It would be even better if we can property handle Accept header, however, I don't think it is mandatory.
* c.f. http://programmers.stackexchange.com/a/170409

If we handle the header, it would require:
* interpret media range and quality factor
* split `Marshaler` into `Marshaler` and `Unmarshaler`.
  * Independent registry for each

**Comment by tamird** [comment]
Rebased.

> I don't think this is a "proper" handling of Accept header since it ignores media range and quality factor of the header. https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.1

Yeah, it's probably incomplete, but still better than the current handling which conflates "content-type" (which specifies the incoming content) with "accept" (which specifies the outgoing content).

> Accept header is not the right way to identify MIME type of request body.

This is not what this PR does. It uses "content-type" to identify the MIME type of the incoming request, but uses "accept" to determine the MIME type that should be used to encode the outgoing response.

> If we handle the header, it would require:

> - interpret media range and quality factor
>   - split Marshaler into Marshaler and Unmarshaler.
>       - Independent registry for each

Yes, we could do these things, but it doesn't seem "required" to me. I'm not sure splitting Marshaler and Unmarshaler is even worth it, but I'll leave that decision to you. Nevertheless, I don't think that refactoring should block this change.

File: runtime/marshaler_registry.go [hunk 1]
```diff
@@ -5,132 +5,87 @@ import (
...
+			break
 		}
-		if inbound != nil && outbound != nil {
-			// Got them both, return
-			return inbound, outbound
+	}
+
+	for _, contentTypeVal := range r.Header[contentTypeHeader] {
+		if m, ok := mux.marshalers.mimeMap[contentTypeVal]; ok {
+			outbound = m
```
Comment (yugui) [inline runtime/marshaler_registry.go]: maybe you mean `inbound = m`?

**Comment by yugui** [comment]
@tamird 
Thank you.  I confused something in the original commits.

I'll add some comments.

File: runtime/marshaler_registry.go [hunk 2]
```diff
@@ -5,132 +5,87 @@ import (
...
 		}
 	}
+
 	if inbound == nil {
-		inbound = defaultMarshaler
+		inbound = mux.marshalers.mimeMap[MIMEWildcard]
 	}
 	if outbound == nil {
-		outbound = defaultMarshaler
+		outbound = mux.marshalers.mimeMap[MIMEWildcard]
```
Comment (yugui) [inline runtime/marshaler_registry.go]: Could you try falling back to `mux.marshalers.mimeMap[contentTypeHeader]` before `MIMEWildcard`?

File: runtime/marshaler_registry.go [hunk 3]
```diff
@@ -5,132 +5,87 @@ import (
...
 
-// marshalerRegistry keeps a mapping from MIME types to mimeMarshalers.
-type marshalerRegistry map[string]*mimeMarshaler
-
-type mimeMarshaler struct {
-	inbound  Marshaler
-	outbound Marshaler
+// marshalerRegistry is a mapping from MIME types to Marshalers.
+type marshalerRegistry struct {
+	mimeMap map[string]Marshaler
```
Comment (yugui) [inline runtime/marshaler_registry.go]: unnecessary change.

File: runtime/marshaler_registry.go [hunk 4]
```diff
@@ -5,132 +5,87 @@ import (
...
-		m := mux.marshalers.lookup(val)
-		if m != nil {
-			if inbound == nil {
-				inbound = m.inbound
-			}
-			if outbound == nil {
-				outbound = m.outbound
-			}
+	for _, acceptVal := range r.Header[acceptHeader] {
+		if m, ok := mux.marshalers.mimeMap[acceptVal]; ok {
```
Comment (yugui) [inline runtime/marshaler_registry.go]: I prefer to keeping `lookup` method to encapsulate internal structure.

File: runtime/marshaler_registry.go [hunk 5]
```diff
@@ -5,132 +5,87 @@ import (
...
-		return nil
+// MakeMarshalerMIMERegistry returns a new registry of marshalers.
+// It allows for a mapping of case-sensitive Content-Type MIME type string to runtime.Marshaler interfaces.
+//
+// For example, you could allow the client to specify the use of the runtime.JSONPb marshaler
+// with a "applicaton/jsonpb" Content-Type and the use of the runtime.JSONBuiltin marshaler
+// with a "application/json" Content-Type.
+// "*" can be used to match any Content-Type.
+// This can be attached to a ServerMux with the marshaler option.
+func MakeMarshalerMIMERegistry() marshalerRegistry {
```
Comment (yugui) [inline runtime/marshaler_registry.go]: With the new API with `WithMarshalerOption`, you don't need to export this function because the only caller of this function is `NewServeMux`.
Also the function should be named `newMarshalerRegistry` by convention.

File: runtime/marshaler_registry.go [hunk 6]
```diff
@@ -5,132 +5,87 @@ import (
...
+	acceptHeader      = http.CanonicalHeaderKey("Accept")
 	contentTypeHeader = http.CanonicalHeaderKey("Content-Type")
+
+	defaultMarshaler = &JSONPb{OrigName: true}
 )
 
 // MarshalerForRequest returns the inbound/outbound marshalers for this request.
 // It checks the registry on the ServeMux for the MIME type set by the Content-Type header.
 // If it isn't set (or the request Content-Type is empty), checks for "*".
+// If that isn't set, uses the ServerMux's InboundMarshaler/OutboundMarshaler.
```
Comment (yugui) [inline runtime/marshaler_registry.go]: remove this line.

File: runtime/marshaler_registry_test.go [hunk 7]
```diff
@@ -34,27 +34,27 @@ func TestMarshalerForRequest(t *testing.T) {
 		wantOut runtime.Marshaler
 	}{
 		{
```
Comment (yugui) [inline runtime/marshaler_registry_test.go]: Could you add scenario for `Accept` header?

Comment (tamird) [inline runtime/marshaler_registry_test.go]: Done.

File: runtime/marshaler_registry.go (hunk 5 above)
Comment (tamird) [inline runtime/marshaler_registry.go]: Unexported. I left it `make` because it returns a value, not a pointer.

File: runtime/marshaler_registry.go (hunk 3 above)
Comment (tamird) [inline runtime/marshaler_registry.go]: what's unnecessary? there's no longer a mapping to two marshalers.

File: runtime/marshaler_registry.go (hunk 2 above)
Comment (tamird) [inline runtime/marshaler_registry.go]: Done.

File: runtime/marshaler_registry.go (hunk 1 above)
Comment (tamird) [inline runtime/marshaler_registry.go]: Yep, done.

File: runtime/marshaler_registry.go (hunk 4 above)
Comment (tamird) [inline runtime/marshaler_registry.go]: none of this is exported and the tuple form of map lookup is more correct.

File: runtime/marshaler_registry.go (hunk 6 above)
Comment (tamird) [inline runtime/marshaler_registry.go]: Done.

**Comment by tamird** [comment]
@yugui ping, could you take another look?

**Comment by yugui** [comment]
LGTM

**Comment by tamird** [comment]
Thanks!
